pkg/raft/tracker/progress.go

```diff
@@ -320,7 +344,26 @@ func (pr *Progress) ShouldSendMsgApp(last, commit uint64) bool {
 		//	- our commit index exceeds the in-flight commit index, and
 		//	- sending it can commit at least one of the follower's entries
 		//	  (including the ones still in flight to it).
-		return pr.CanBumpCommit(commit)
+		if pr.CanBumpCommit(commit) {
+			return true
+		}
+
+		// Send an empty MsgApp containing the latest commit index if we know that
+		// the follower's commit index is stale and we haven't recently sent a
+		// MsgApp (according to the MsgAppProbesPaused flag).
+
+		// NOTE: This is a different condition than the one above because we only
+		// send this message if pr.MsgAppProbesPaused is false. After this message,
+		// pr.MsgAppProbesPaused will be set to true until we receive a heartbeat
+		// response from the follower. In contrast, the condition above can keep
+		// sending empty MsgApps until we have sent the latest commit index to the
+		// follower.
+		// TODO(iskettaneh): Remove the dependency on MsgAppProbesPaused to send
+		// MsgApps.
+		if advanceCommit {
+			return pr.IsFollowerCommitStale(commit) && !pr.MsgAppProbesPaused
```

Comment:
Consider inlining `IsFollowerCommitStale`, analogous to `if pr.Match < last && !paused` check above. Or don't do it if you intend to make the condition in it more complex (like check `pr.Next` too, see the corresponding comment above).